### PR TITLE
CHUNGUS

### PR DESCRIPTION
**Pull request by ciebson123**
# Description
Faster CHUNGUS implementation
Works with:
- fsdp
- gradient accumulation
- activation checkpointing
- switch (auxliary loss)

Equivalence comparison: https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9b61a2d0-db32-45c6-9b8f-20075f497388&detailsTab=charts&shortId=LLMRANDOM-7701&dash=charts&compare=OwJgnAHANArNCMVHOMp7EGY0mEoA&chartFilter=loss_interval%2F100

All look close

Speed comparison: https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9b61a2d0-db32-45c6-9b8f-20075f497388&detailsTab=charts&shortId=LLMRANDOM-7701&dash=charts&chartFilter=step&compare=OwNgzCA0CMM0A

New Chungus faster than old

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [ ] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [ ] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [ ] Are all functions placed in appropriate files?
   - [ ] Is the PR name meaningful and descriptive?
   - [ ] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/utils/conditional_trainer.py [hunk 1]
```diff
@@ -185,16 +183,15 @@ def _train_step(
             self._log_auxiliary_losses(aux_info["losses"], step)
         self._save_weights(step)
 
-    def calculate_loss_and_maybe_optimize(
-        self, processed_batch: LLMBatch, should_optimize: bool
-    ):
+    def calculate_loss_and_maybe_optimize(self, processed_batch: LLMBatch):
```
Comment (Krystian-Krol) [inline research/conditional/utils/conditional_trainer.py:186]: I wont review the code since I have no idea about this part, but is `_and_maybe_optimize` part of the name still true?

Comment (ciebson123) [reply research/conditional/utils/conditional_trainer.py:186]: Yes, it optimizes only in training

File: research/conditional/utils/model_utils.py [hunk 2]
```diff
@@ -156,40 +196,54 @@ def calculate_llm_loss(
...
+        return loss, aux_info
+
+    loss, aux_info = hack_for_python_garbage_collection()
+    for key, value in aux_info["losses"].items():
+        aux_info["losses"][key] = value / num_checkpoint_accumulation_steps
+    if model.training:
+        loss_to_optimize = loss.clone()
+        for value in aux_info["losses"].values():
+            loss_to_optimize += value
+        backward_maybe_with_scaler(loss_to_optimize, mixed_precision_dtype, scaler)
```
Comment (crewtool) [inline research/conditional/utils/model_utils.py:243]: After discussion: rename to `run_backward`

File: research/conditional/utils/model_utils.py [hunk 3]
```diff
@@ -156,40 +196,54 @@ def calculate_llm_loss(
...
+        with torch.autocast(
+            device_type="cuda", enabled=mixed_precision, dtype=mixed_precision_dtype
+        ):
+            model_output = model(input_tokens)
 
-    with torch.autocast(
-        device_type="cuda", enabled=mixed_precision, dtype=mixed_precision_dtype
-    ):
-        model_output = model(input_tokens)
+        # move the gt tokens and mask to the same device as the model output - they should be on the same device for loss calculation
```
Comment (crewtool) [inline research/conditional/utils/model_utils.py:212]: Is it for fsdp? Do we need to make sure that they are on the same device by ourselves?

File: research/conditional/utils/model_utils.py [hunk 4]
```diff
@@ -156,40 +196,54 @@ def calculate_llm_loss(
...
+        gt_tokens = gt_tokens.to(model_output.device)
+        mask = mask.to(model_output.device)
 
-    # move the gt tokens and mask to the same device as the model output - they should be on the same device for loss calculation
-    gt_tokens = gt_tokens.to(model_output.device)
-    mask = mask.to(model_output.device)
+        mask_loss = F.cross_entropy(
+            model_output.reshape(-1, vocab_size),
```
Comment (crewtool) [inline research/conditional/utils/model_utils.py:217]:
```suggestion
            model_output.reshape(-1, model_output.shape[-1]),
```

File: research/conditional/utils/model_utils.py [hunk 5]
```diff
@@ -68,76 +67,106 @@
...
         total_masked_tokens = 0
-        for chunged_input, chunged_gt, chunged_mask in zip(
-            chunged_inputs, chunged_non_masked_inputs, chunged_non_masked_masks
+        for chunged_encoder_output, chunged_gt, chunged_mask in zip(
+            chunged_encoder_outputs, chunged_non_masked_inputs, chunged_non_masked_masks
         ):
             (
                 partial_loss_output,
                 partial_correct_tokens,
                 partial_masked_tokens,
```
Comment (crewtool) [inline research/conditional/utils/model_utils.py:152]: How about `single_chung_*?

Comment (crewtool) [reply research/conditional/utils/model_utils.py:152]: and ommit `output` suffix

File: research/conditional/utils/model_utils.py [hunk 6]
```diff
@@ -68,76 +67,106 @@
...
+        correct_tokens = gt.long() == output.argmax(dim=-1)
+        correct_tokens = correct_tokens.long().reshape(-1) * mask.reshape(-1)
+        correct_tokens = correct_tokens.sum()
+
+        total_tokens = mask.sum()
+
+    return loss[mask.reshape(-1) == 1], correct_tokens, total_tokens
+
+
+def backward_maybe_with_scaler(
```
Comment (crewtool) [inline research/conditional/utils/model_utils.py:105]: `rub_backward`

**Review by crewtool — CHANGES_REQUESTED** [review]
Excellent job!
I left some minor comments.
Could you also add test for  num_checkpoint_accumulation_steps != 1?

File: research/conditional/utils/model_utils.py [hunk 7]
```diff
@@ -68,76 +67,106 @@
...
+            chunged_encoder_outputs, chunged_non_masked_inputs, chunged_non_masked_masks
         ):
             (
                 partial_loss_output,
                 partial_correct_tokens,
                 partial_masked_tokens,
-            ) = checkpoint(
-                make_custom_forward(), chunged_input, chunged_gt, chunged_mask
+            ) = calculate_single_chung_loss(
+                model,
```
[the remark this answers is not in the document]
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:154]: True, good idea

File: research/conditional/utils/model_utils.py (hunk 3 above)
Comment (ciebson123) [reply research/conditional/utils/model_utils.py:212]: I think so, at least it was here in the previous implementation

File: research/conditional/utils/conditional_trainer.py [hunk 8]
```diff
@@ -207,61 +204,42 @@ def calculate_loss_and_maybe_optimize(
...
+            total_cross_entropy_loss += cross_entropy_loss
             correct_tokens_value += aux_info["correct_tokens"]
             total_masked_tokens_value += aux_info["total_masked_tokens"]
 
             for key, value in aux_info["losses"].items():
-                losses[key] = losses.get(key, 0) + value
+                losses[key] = losses.get(key, 0) + value.item()
+
+        if self.model.training:
+            self._apply_gradient()
```
[the remark this answers is not in the document]
Comment (ciebson123) [reply research/conditional/utils/conditional_trainer.py:219]: True, I can change that, but it was around here in the previous implementation.

[the remark this answers is not in the document]
Comment (ciebson123) [reply research/conditional/utils/conditional_trainer.py:219]: Done

**Review by crewtool — APPROVED** [review]
LGTM
